File: SymmetricDDA.py
from BaseClass import *


class SymmetricDDA(Line):
    def __init__(self,x1, y1, x2, y2,name):
        super(SymmetricDDA, self).__init__(x1, y1, x2, y2 ,name)

    def build_N(self):
        # 生成一个合理N 用来分隔线段 通过移位操作就可以实现
        self.dx = self.x2-self.x1
        self.dy = self.y2-self.y1
        max_increment = max(self.dx,self.dy)
        n = 0
        while 1:
            max_increment = max_increment >> 1
            n += 1
            if max_increment < 1:
                break
        self.N = 2**n

    def display(self):
        self.build_N()
        x_increment = self.dx/self.N
        y_increment = self.dy/self.N
        glBegin(GL_POINTS)
        glVertex2d(int(self.x1), int(self.y1))
        x,y = self.x1, self.y1
        for i in range(self.N):
                x, y = x + x_increment, y + y_increment
                glVertex2d(int(x), int(y))
        glEnd()
        glFlush()

if __name__ == '__main__':
    line = SymmetricDDA(0,0, 280, 480, 'SymmetricDDA.py')
    line.run()

File: cpp/SymmetricDDA.cpp
#define GLUT_DISABLE_ATEXIT_HACK
#include<windows.h>
#include<gl/gl.h> 
#include<gl/glut.h>
#include<math.h>
using namespace std;
void init()
        { 
        glutInitDisplayMode(GLUT_SINGLE | GLUT_RGB);
        glutInitWindowSize(800, 600);
        glutInitWindowPosition(200, 200);
        glutCreateWindow("line");
		}
		
		
		
    void end_init()
		{
        glClearColor(1.0, 1.0, 1.0, 0);
        glColor3f(0, 0, 0);
        gluOrtho2D(0.0, 500.0, 0.0, 500.0);
        glutMainLoop();
		}
		
		
		


    void display()
	{
		int x1=0;int y1=0;int x2=180;int y2=180;
		float dx,dy;
		dx = x2-x1;
        dy = y2-y1;
        int max_increment = (dx>dy)?dx:dy;
        int n = 0;float N;
        while (1)
		{
            max_increment = max_increment >> 1;
            n ++;
            if (max_increment < 1) break;
        	N = pow(2,n);
		}
		float x_increment = dx/N;
        float y_increment = dy/N;
        glBegin(GL_POINTS);
        glVertex2d(int(x1), int(y1));
        float x,y; x = x1; y= y1;
        for (int i=0;i<N;i++)
				{
                x  = x + x_increment;  y= y + y_increment;
                glVertex2d(int(x), int(y));
                }
        glEnd();
        glFlush();

	}
    	
    void run()
	    {
        glutDisplayFunc(display);
        end_init();
		}

int main(int argc, char** argv)

{
    glutInit(&argc, argv);
	init();
	run();
	end_init();
	return 0;
}


